File: services/ingestion/health.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"sync/atomic"

	"github.com/prometheus/client_golang/prometheus/promhttp"
)

// startHealthServer exposes /live (process is up — never fails on a
// dependency), /ready (this instance can currently do useful work), and
// /metrics (Prometheus scrape endpoint). Liveness deliberately does not
// depend on MQTT/Kafka health: a transient broker outage should not cause
// an orchestrator to kill and restart a perfectly healthy process.
func startHealthServer(port string, mqttConnected *atomic.Bool, sink *kafkaSink) {
	mux := http.NewServeMux()

	mux.HandleFunc("/live", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("ok"))
	})

	mux.HandleFunc("/ready", func(w http.ResponseWriter, r *http.Request) {
		state := sink.breakerState()
		ready := mqttConnected.Load() && state != "OPEN"

		w.Header().Set("Content-Type", "application/json")
		if !ready {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"ready":                 ready,
			"mqtt_connected":        mqttConnected.Load(),
			"kafka_circuit_breaker": state,
		})
	})

	mux.HandleFunc("/health", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"mqtt_connected":        mqttConnected.Load(),
			"kafka_circuit_breaker": sink.breakerState(),
		})
	})

	mux.Handle("/metrics", promhttp.Handler())

	go func() {
		// ListenAndServe only returns once the listener stops; nothing in
		// this service ever calls Shutdown/Close on it, so any return here
		// is unexpected (most likely a bind failure, e.g. the port already
		// in use) and worth surfacing loudly rather than leaving /live and
		// /ready silently unreachable for the rest of the process's life.
		if err := http.ListenAndServe(":"+port, mux); err != nil { //nolint:gosec // internal-only health/metrics endpoint
			log.Printf("ingestion: health/metrics server on :%s stopped: %v", port, err)
		}
	}()
}

File: services/ingestion/health.py
"""HTTP health/metrics server, the Python port of health.go. Same three
endpoints, same semantics: /live never depends on a dependency (a
transient broker outage should not cause an orchestrator to kill and
restart a perfectly healthy process), /ready reflects real MQTT+Kafka
state, /metrics is the Prometheus scrape endpoint.
"""

from __future__ import annotations

import json
import logging
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from prometheus_client import CONTENT_TYPE_LATEST, generate_latest

from kafka_producer import KafkaSink
from mqtt_client import AtomicBool

_logger = logging.getLogger("ingestion")


def _make_handler(mqtt_connected: AtomicBool, sink: KafkaSink) -> type[BaseHTTPRequestHandler]:
    class Handler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *args) -> None:  # noqa: A002 - silence stdlib's default access log
            pass

        def do_GET(self) -> None:
            if self.path == "/live":
                self.send_response(200)
                self.end_headers()
                self.wfile.write(b"ok")
                return

            if self.path == "/ready":
                state = sink.breaker_state()
                connected = mqtt_connected.get()
                ready = connected and state != "OPEN"
                body = json.dumps(
                    {"ready": ready, "mqtt_connected": connected, "kafka_circuit_breaker": state}
                ).encode("utf-8")
                self.send_response(200 if ready else 503)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(body)
                return

            if self.path == "/health":
                body = json.dumps(
                    {"mqtt_connected": mqtt_connected.get(), "kafka_circuit_breaker": sink.breaker_state()}
                ).encode("utf-8")
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(body)
                return

            if self.path == "/metrics":
                body = generate_latest()
                self.send_response(200)
                self.send_header("Content-Type", CONTENT_TYPE_LATEST)
                self.end_headers()
                self.wfile.write(body)
                return

            self.send_response(404)
            self.end_headers()

    return Handler


def start_health_server(port: str, mqtt_connected: AtomicBool, sink: KafkaSink) -> ThreadingHTTPServer:
    server = ThreadingHTTPServer(("", int(port)), _make_handler(mqtt_connected, sink))
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server
